xmppserver/src/main/java/org/jivesoftware/openfire/net/SocketUtil.java

```diff
@@ -60,59 +129,132 @@ public static Map.Entry<Socket, Boolean> createSocketToXmppDomain( String xmppDo
+            Instant nextJobNotBefore = Instant.EPOCH;
 
-            try
-            {
-                // (re)initialize the socket.
-                socket = new Socket();
+            Log.debug( "Use DNS to resolve remote hosts for the provided XMPP domain '{}' (default port: {}) ...", xmppDomain, port );
+            final List<Set<DNSUtil.WeightedHostAddress>> remoteHosts = DNSUtil.resolveXMPPDomain(xmppDomain, port);
+            for (final Set<DNSUtil.WeightedHostAddress> prioritySet : remoteHosts) {
+                if (executor.isTerminating() || executor.isTerminated() || executor.isShutdown()) {
+                    Log.trace("Aborting resolution, as the executor is being shut down (likely cause: we successfully identified a result).");
+                    return;
+                }
+                final boolean preferIpv4 = InetAddress.getLoopbackAddress() instanceof Inet4Address; // Follow the preference of the JVM.
+                final HappyEyeballsResolver resolver = new HappyEyeballsResolver(new LinkedList<>(prioritySet), preferIpv4, RESOLUTION_DELAY.getValue());
+                try {
+                    resolver.start();
+                    while (!resolver.isDone() && !executor.isTerminating() && !executor.isTerminated() && !executor.isShutdown() && Instant.now().isBefore(deadline)) {
+                        final HappyEyeballsResolver.XmppServiceAddress resolvedAddress = resolver.getNext(); // Blocks.
+                        Log.trace("Next resolved address for '{}': {}", xmppDomain, resolvedAddress);
+                        if (resolvedAddress == null) {
+                            continue;
+                        }
 
-                Log.debug( "Trying to create socket connection to XMPP domain '{}' using remote host: {}:{} (blocks up to {} ms) ...", xmppDomain, realHostname, realPort, socketTimeout );
-                socket.connect( new InetSocketAddress( realHostname, realPort ), socketTimeout );
-                Log.debug( "Successfully created socket connection to XMPP domain '{}' using remote host: {}:{}!", xmppDomain, realHostname, realPort );
-                return new AbstractMap.SimpleEntry<>(socket, directTLS);
-            }
-            catch ( Exception e )
-            {
-                Log.debug( "An exception occurred while trying to create a socket connection to XMPP domain '{}' using remote host {}:{}", xmppDomain, realHostname, realPort, e );
-                Log.warn( "Unable to create a socket connection to XMPP domain '{}' using remote host: {}:{}. Cause: {} (a full stacktrace is logged on debug level)", xmppDomain, realHostname, realPort, e.getMessage() );
-                try
-                {
-                    if ( socket != null )
-                    {
-                        socket.close();
-                        socket = null;
+                        if (!JiveGlobals.getBooleanProperty(ConnectionSettings.Server.ENABLE_OLD_SSLPORT, true) && resolvedAddress.isDirectTLS()) {
```

Comment:
Couple of old-style settings here might be worth considering migrating while you're in the area - this line plus 5 lines later.

Reply:
I tried that, but it quickly spiraled out of control. Those same settings are used elsewhere, and when you start migrating things in `ConnectionSettings` you really aught to do all of the settings in there. In the end, I reverted and created https://igniterealtime.atlassian.net/browse/OF-2888 (that I hope you will self-assign ;) ) to tackle this in a separate effort.